clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -0,0 +1,191 @@
+      byte[] requestBody) {
+    String[] requestParts = RequestHelper.getRequestParts(requestPath);
+
+    ManagedChannel channel = getChannel(requestPath);
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setStoreName(requestParts[2])
+        .setResourceName(requestParts[2])
+        .setPartition(-1)
+        .setKeyBytes(ByteString.copyFrom(requestBody))
+        .build();
+
+    VeniceReadServiceGrpc.VeniceReadServiceStub clientStub = VeniceReadServiceGrpc.newStub(channel);
+    GrpcTransportClientBatchCallback callback = new GrpcTransportClientBatchCallback(clientStub, request);
+
+    return callback.post();
+  }
+
+  @Override
+  public void close() throws IOException {
+    for (Map.Entry<String, ManagedChannel> entry: serverGrpcChannels.entrySet()) {
+      entry.getValue().shutdown();
+    }
+  }
+
+  public static class GrpcTransportClientCallback {
+    // start exception handling
+    private final CompletableFuture<TransportClientResponse> valueFuture;
+    private final VeniceReadServiceGrpc.VeniceReadServiceStub clientStub;
+    private final VeniceClientRequest request;
+
+    public GrpcTransportClientCallback(
+        VeniceReadServiceGrpc.VeniceReadServiceStub clientStub,
+        VeniceClientRequest request) {
+      this.clientStub = clientStub;
+      this.request = request;
+      this.valueFuture = new CompletableFuture<>();
+    }
+
+    public CompletableFuture<TransportClientResponse> get() {
+      clientStub.get(request, new StreamObserver<VeniceServerResponse>() {
+        @Override
+        public void onNext(VeniceServerResponse value) {
+          valueFuture.complete(
+              new TransportClientResponse(
+                  value.getSchemaId(),
+                  CompressionStrategy.NO_OP,
+                  value.getData().toByteArray()));
+        }
+
+        @Override
+        public void onError(Throwable t) {
+          LOGGER.error("Error in gRPC request", t);
+          valueFuture.completeExceptionally(t);
+        }
+
+        @Override
+        public void onCompleted() {
+          LOGGER.info("Completed gRPC request");
+        }
+      });
+
+      return valueFuture;
+    }
+
+  }
+
+  public static class GrpcTransportClientBatchCallback {
+    private final CompletableFuture<TransportClientResponse> valueFuture;
+    private final VeniceReadServiceGrpc.VeniceReadServiceStub clientStub;
+    private final VeniceClientRequest request;
+
+    public GrpcTransportClientBatchCallback(
+        VeniceReadServiceGrpc.VeniceReadServiceStub clientStub,
+        VeniceClientRequest request) {
+      this.clientStub = clientStub;
+      this.request = request;
+      this.valueFuture = new CompletableFuture<>();
+    }
+
+    public CompletableFuture<TransportClientResponse> post() {
+      clientStub.batchGet(request, new StreamObserver<VeniceServerResponse>() {
+        @Override
+        public void onNext(VeniceServerResponse value) {
```

Comment:
Can this function be invoked multiple times?
Also it seems it doesn't support the proper compression header here.

Reply:
post() will be invoked once, and in this implementation onNext will also be invoked once since we are not streaming any of the responses/requests. The compression header has been added.